### PR TITLE
extend `sectionify` to support `page.Route` array

### PR DESCRIPTION
**Pull request by donpark**
`sectionify` function in `lib/route` module currently supports only simple paths with **at most one parameter**, which is matched by value then **discarded**. This results in shortcomings reported in issue #16902 which are Tracks `calypso_page_view` events recorded with:

1. parameter values in `path` property.
2. `path` property without parameter names.

This PR extends `sectionify` function in `lib/route` module to use optional array of `page` modules' `Route` instances so correct parameterized paths can be determined for any known path.

`calypso_page_view` event property `path` being reported **before** this PR:

    /checkout/domain_map:69thclergycouncil.com/renew/5164008
    /checkout/theme:elemin
    /checkout/thank-you/25222194

event property `path` expected to be reported **after** this PR:

    /checkout/:product/renew/:receipt
    /checkout/:product
    /checkout/thank-you/:receipt

Test Plan:

    npm run test-client client/lib/route

Note: Despite `[Size] L` labeling, this PR is actually very small in impact area, specifically limited to checkout controller.

### CODE REVIEW

**Comment by yanirs** [comment]
I haven't tested it, but the code looks good to me. However, do we want to completely strip the products from the path? It may be better to replace paths like `/checkout/domain_map:site.com/renew/1234` with `/checkout/domain_map::domain/renew/:receipt`, so we can still run queries on specific products.

Probably best to run it by someone in marketing, like @Essayoh.

**Comment by jsnajdr** [comment]
Is it possible to make `sectionifyUsingRoutes` a completely independent function and call it as such from the places that need it?

The `sectionify` function is already hard to understand: it is not documented, the name is not very self-explanatory... Overloading the second parameter with two completely unrelated meanings is not an improvement.

Two tips for making the code a bit cleaner:
- use `Array.isArray( x )` to check if `x` is array with 100% accuracy
- use `for (const route of routes)` to iterate over an array

**Comment by donpark** [comment]
@yanirs imo both routes and how we track page view needs a bit of refactoring. I intentionally minimized amount of changes to limit impact areas for now.

Note that:

1. Details you inquired about, parameter values, are extracted but then discarded within `sectionifyUsingRoutes` function, because intermediate layers of functions do not support them. after refactoring we can although we should keep amount of process down to minimum on collection-side.
2. This solution is hack to workaround the fact that `context` being passed could but don't have the original path pattern. To fix this, we'll need to fork `page` module. Again, my minimal change rule applied here.

@jsnajdr No because that would increase the surface area. Only `sectionify` was being used broadly and just extending it makes more sense than exposing a similar function externally. I agree that `sectionify` is hard to understand and very likely misnamed. I disagree this is the right time nor right context to correct this. I think routing and event tracking will soon need a refactor and doubt `sectionify` will survive it.

Re your tips, thx. I'm still unclear on what Calypso's target ES version is so I went with conservative options.

**Comment by donpark** [comment]
Changes I made with latest push:

1. Restored original `sectionify`.
2. Exposed `sectionifyWithRoutes` to be invoked directly when route may be complex.
3. Report path parameters as event properties

File: client/lib/analytics/index.js [hunk 1]
```diff
@@ -282,12 +282,17 @@ const analytics = {
...
-		recordPageView: function( urlPath ) {
+		recordPageView: function( urlPath, params ) {
 			let eventProperties = {
 				path: urlPath,
 				do_not_track: doNotTrack() ? 1 : 0
 			};
 
+			// add optional path params
+			if ( params ) {
+				eventProperties = assign( eventProperties, params );
```
Comment (jsnajdr) [inline client/lib/analytics/index.js]: `lodash.assign` mutates the `eventProperties` object. Assigning the return value to `eventProperties` is not necessary. But also not harmful :)

Comment (donpark) [reply client/lib/analytics/index.js]: thx. I tend to keep redundant assignment for readability sake.

**Pull request merged**